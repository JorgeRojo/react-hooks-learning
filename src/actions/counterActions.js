import {
  COUNTER_DECREASE,
  COUNTER_INCREASE,
  COUNTER_RESET,
  COUNTER_SET,
} from './actionTypes';

export const counterDecrease = () => ({
  type: COUNTER_DECREASE,
});

export const counterIncrease = () => ({
  type: COUNTER_INCREASE,
});

export const counterReset = () => ({
  type: COUNTER_RESET,
});

export const counterSet = number => ({
  type: COUNTER_SET,
  payload: { number },
});
