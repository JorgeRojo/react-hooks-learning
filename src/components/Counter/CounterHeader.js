import React from 'react';
import { connect } from 'react-redux';
import { selectCounter } from '../../reducer';

const CounterHeader = ({ counter }) => {
  return <h1>Counter: {counter}</h1>;
};

const mapStateToProps = state => ({
  counter: selectCounter(state),
});

export default connect(mapStateToProps)(CounterHeader);
