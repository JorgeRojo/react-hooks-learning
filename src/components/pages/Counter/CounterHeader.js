import React from 'react';
import { connect } from 'react-redux';
import { selectCounter } from '../../../reducer';

const CounterHeader = ({ counter }) => {
  return <h1 className="display-4 mb-4">Counter: {counter}</h1>;
};

const mapStateToProps = state => ({
  counter: selectCounter(state),
});

export default connect(mapStateToProps)(CounterHeader);
