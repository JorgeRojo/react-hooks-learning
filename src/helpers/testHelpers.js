import React from 'react';

export const mockComponent = (name, props) => (
  <code name={name}>{JSON.stringify(props)}</code>
);
