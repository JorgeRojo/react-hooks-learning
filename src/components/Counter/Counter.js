import React, { useCallback, useState, useEffect } from "react";
import { useSelector, shallowEqual } from "react-redux";
import { withReduxDispatch } from "../../store";

import * as counterActions from "../../actions/counterActions";
import { selectCounter } from "../../reducer";

import { getNumberFromEvent } from "./counterHelpers";

const {
  counterIncrease,
  counterDecrease,
  counterSet,
  counterReset
} = withReduxDispatch(counterActions);

const Counter = () => {
  const counter = useSelector(selectCounter, shallowEqual);

  const [number, setNumber] = useState(counter);

  useEffect(() => {
    setNumber(counter);
  }, [counter]);

  const handleClickIncrease = useCallback(() => {
    counterIncrease();
  }, []);

  const handleClickDecrease = useCallback(() => {
    counterDecrease();
  }, []);

  const handleClickReset = useCallback(() => {
    counterReset();
  }, []);

  const handleChangeNumber = useCallback(event => {
    setNumber(getNumberFromEvent(event));
  }, []);

  const handleClickSet = useCallback(() => {
    counterSet(number);
  }, [number]);

  const handleKeyPressSet = useCallback(
    ({ key }) => {
      if (key === "Enter") {
        handleClickSet();
      }
    },
    [handleClickSet]
  );

  const handleChangeSet = useCallback(event => {
    counterSet(getNumberFromEvent(event));
  }, []);

  return (
    <>
      <h2>Counter: {counter}</h2>
      <hr />
      <button onClick={handleClickIncrease}>+ 1</button>

      <span> | </span>
      <button onClick={handleClickDecrease}>- 1</button>

      <span> | </span>
      <button onClick={handleClickReset}>= 0</button>

      <span> | </span>
      <button onClick={handleClickSet}> = </button>
      <input
        value={number}
        onKeyPress={handleKeyPressSet}
        onChange={handleChangeNumber}
        type="text"
      />

      <span> | </span>
      <input value={counter} onChange={handleChangeSet} type="text" />
    </>
  );
};

export default Counter;
