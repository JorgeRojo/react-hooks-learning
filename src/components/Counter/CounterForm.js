import React, { useCallback, useState, useEffect } from "react";
import { useSelector, shallowEqual } from "react-redux";
import { withReduxDispatch } from "../../store";

import * as counterActions from "../../actions/counterActions";
import { selectCounter } from "../../reducer";

import { getNumberFromEvent } from "./counterHelpers";

const {
  counterIncrease,
  counterDecrease,
  counterSet,
  counterReset
} = withReduxDispatch(counterActions);

const CounterForm = () => {
  const counter = useSelector(selectCounter, shallowEqual);

  const [number, setNumber] = useState(counter);

  useEffect(() => {
    setNumber(counter);
  }, [counter]);

  const handleChangeNumber = useCallback(event => {
    setNumber(getNumberFromEvent(event));
  }, []);

  const handleClickSet = useCallback(() => {
    counterSet(number);
  }, [number]);

  const handleKeyPressSet = useCallback(
    ({ key }) => {
      if (key === "Enter") {
        counterSet(number);
      }
    },
    [number]
  );

  const handleChangeSet = useCallback(event => {
    counterSet(getNumberFromEvent(event));
  }, []);

  return (
    <>
      <p>
        <button aria-label="counterIncreaseBtn" onClick={counterIncrease}>
          + 1
        </button>

        <span> | </span>
        <button aria-label="counterDecreaseBtn" onClick={counterDecrease}>
          - 1
        </button>

        <span> | </span>
        <button aria-label="counterResetBtn" onClick={counterReset}>
          = 0
        </button>
      </p>

      <p>
        <input
          aria-label="handleChangeNumberInput"
          value={number}
          onKeyPress={handleKeyPressSet}
          onChange={handleChangeNumber}
          type="text"
        />
        <button aria-label="handleClickSetBtn" onClick={handleClickSet}>
          SET
        </button>
      </p>

      <p>
        <label htmlFor="number-live"> counter = </label>
        <input
          aria-label="handleChangeSetInput"
          id="number-live"
          value={counter}
          onChange={handleChangeSet}
          type="text"
        />
      </p>
    </>
  );
};

export default CounterForm;
