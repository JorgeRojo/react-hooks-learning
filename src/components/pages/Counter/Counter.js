import React, { useEffect } from 'react';
import CounterHeader from './CounterHeader';
import CounterForm from './CounterForm/CounterForm';

import { reduxActionsDispatchers } from '../../../store';
import { counterReset } from '../../../actions/counterActions';
const { dispatchCounterReset } = reduxActionsDispatchers({ counterReset });

const Counter = () => {
  useEffect(() => dispatchCounterReset);

  return (
    <div className="container text-center mt-4">
      <CounterHeader />
      <CounterForm />
    </div>
  );
};

export default Counter;
