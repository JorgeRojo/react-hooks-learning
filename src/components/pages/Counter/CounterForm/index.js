import CounterForm from './CounterForm';
export default CounterForm;
