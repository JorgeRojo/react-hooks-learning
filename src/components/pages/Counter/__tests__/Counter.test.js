import React from 'react';
import ReactDOM from 'react-dom';
import { render, cleanup } from '@testing-library/react';
import { Provider } from 'react-redux';

import store from '../../../../store';
import { counterReset } from '../../../../actions/counterActions';

import Counter from '../Counter';
import { mockComponent } from '../../../../helpers/testHelpers';

jest.mock('../../../../store', () =>
  require('../../../../store/__mocks__/mockStore')
);
jest.mock('../../../../actions/counterActions');
jest.mock('../CounterHeader', () => props =>
  mockComponent('CounterHeader', props)
);
jest.mock('../CounterForm/CounterForm', () => props =>
  mockComponent('CounterForm', props)
);

describe('Counter Component', () => {
  const component = (
    <Provider store={store}>
      <Counter />
    </Provider>
  );

  beforeEach(() => {
    counterReset.mockImplementation(() => ({ type: 'counterReset' }));
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the component', () => {
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });

  it('counterReset called on component unmount', () => {
    const { unmount } = render(component);
    unmount(component);
    expect(counterReset).toHaveBeenCalledTimes(1);
  });
});
