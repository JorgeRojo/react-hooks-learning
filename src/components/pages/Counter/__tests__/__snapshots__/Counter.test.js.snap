// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component renders the component 1`] = `
<div>
  <div
    class="container text-center mt-4"
  >
    <code
      name="CounterHeader"
    >
      {}
    </code>
    <code
      name="CounterForm"
    >
      {}
    </code>
  </div>
</div>
`;
